python/paddle/fluid/contrib/slim/quantization/qat2_int8_mkldnn_pass.py

```diff
@@ -117,20 +129,46 @@ def _gather_scales(self, graph):
 
                 input_name = op.input("X")[0]
                 scale_name = op.input("InScale")[0]
+                output_name = op.output("Out")[0]
                 # Gather new weights scale after folding batchnorm in convolution
                 scale = np.array(1.0 / self._load_param(
                     self._scope, scale_name)[0]).astype(np.float64)
                 lod_tensor = self._convert_scale2tensor(scale)
                 use_unsigned_int = False
-                self._var_quant_scales[input_name] = (use_unsigned_int,
-                                                      lod_tensor)
-                self._var_quant_scales[scale_name.replace(".scale", "")] = (
-                    use_unsigned_int, lod_tensor)
+                _add_scale_for_vars([input_name, output_name], use_unsigned_int,
+                                    lod_tensor)
+
+        return graph
 
+    def _gather_weight_scales_from_fake(self, graph):
+        for op in graph.all_op_nodes():
             if op.name() in self._fake_dequantize_types:
                 input_name = op.input("X")[0]
-                _max_range = op.op().attr("max_range")
-                self._weight_scales[input_name] = _max_range
+                if op.op().has_attr("max_range"):
+                    _max_range = np.array(op.op().attr("max_range")).astype(
+                        np.float64)
+                    self._weight_scales[input_name] = _max_range
+
+        return graph
+
+    def _gather_scales_from_attr(self, graph):
+        def _get_attr_scale_from_op(op, op_types, op_out_name):
+            if op.name() in op_types and op.op().has_attr("out_threshold"):
+                assert op_out_name in op.op().outputs(
+                ), "Operator {} does not have the output {}.".format(
+                    op.name(), op_out_name)
+                attr_scale = op.op().attr("out_threshold")
+                scale = np.array(1.0 / attr_scale).astype(np.float64)
+                scale_lod_tensor = self._convert_scale2tensor(scale)
+                use_unsigned_int = False
+                out_var_name = op.output(op_out_name)[0]
+                self._var_quant_scales[out_var_name] = (use_unsigned_int,
+                                                        scale_lod_tensor)
+
+        for op in graph.all_op_nodes():
+            _get_attr_scale_from_op(op, ['batch_norm'], 'Y')
+            _get_attr_scale_from_op(op, ['relu'], 'Out')
```

Comment:
It should gather all output scales for all op nodes that have out_threshold in attribute.

For example,  a quantized conv2d is not followed by batch_norm and relu op, we need gather the output scales from conv2d attribute.

Reply:
Done.